Instalando dependencias

# 03 Embeddings y VectorStore

## Información

### Objetivo

Este notebook implementa la tercera etapa del pipeline RAG: la generación de embeddings y la preparación de la información para su almacenamiento en una base vectorial.

Su propósito es transformar cada fragmento (Chunk) en una representación vectorial que permita realizar búsquedas por similitud semántica durante la etapa de recuperación de información.

### Objetivos específicos
Cargar un modelo de embeddings basado en Sentence Transformers.
Generar la representación vectorial de los chunks creados en la etapa anterior.
Comprender la estructura y dimensiones de los embeddings generados.
Asociar cada embedding con su chunk correspondiente.
Preparar la información para su posterior almacenamiento en un Vector Store.

### Entradas
Lista de objetos Chunk.
Contenido textual previamente limpiado y fragmentado.
Modelo all-MiniLM-L6-v2.

### Salidas
Chunks enriquecidos con su representación vectorial.
Embeddings listos para ser indexados en una base vectorial.
Validación del correcto funcionamiento del modelo de embeddings.

### Contexto dentro del pipeline RAG

Hasta esta etapa el flujo del proyecto es:

```
PDF
↓
Document
↓
Limpieza
↓
Chunking
↓
Embeddings ← Etapa actual

```


En la siguiente etapa estos embeddings serán almacenados en una base vectorial para recuperar la información más relevante ante una consulta del usuario.

## ¿Qué es un embedding?

Un embedding es una representación numérica de un texto.

Los modelos de embeddings transforman palabras, frases o documentos en vectores de alta dimensión donde textos con significados similares quedan cercanos entre sí.

Esta representación permite que el sistema encuentre información por significado y no únicamente por coincidencia exacta de palabras.

## Modelo seleccionado

Para la versión 1.0 de SOPHIA se utilizará el modelo:

all-MiniLM-L6-v2

Motivos de la elección:

- Modelo liviano y rápido.
- Funciona correctamente en CPU.
- Buen equilibrio entre rendimiento y calidad.
- Ampliamente utilizado en proyectos RAG.
- Adecuado para un MVP orientado a portafolio.

### Importación de librerías

In [30]:
# Esta librería sentence-transformes nos permite descargar el modelo, cargarlo y generar embeddings.
# %pip install sentence-transformers

### Carga del modelo

In [31]:
from sentence_transformers import SentenceTransformer

model = SentenceTransformer("all-MiniLM-L6-v2")

Loading weights: 100%|██████████| 103/103 [00:00<00:00, 3578.46it/s]


### Generación de un embedding de prueba

Como primera validación se generará el embedding de un texto simple para comprobar que el modelo funciona correctamente.

In [32]:
texto = """
La Municipalidad de Valdivia cuenta con un Plan Comunal de Emergencia
para coordinar la respuesta frente a desastres naturales.
"""

In [34]:
embedding = model.encode(texto)

## Inspección del embedding

El resultado es un vector numérico que representa el significado semántico del texto.

No es necesario interpretar individualmente cada valor; lo relevante es que el modelo utiliza este vector para calcular similitudes entre textos.

In [35]:
type(embedding)

numpy.ndarray

In [36]:
print("Dimensiones:", len(embedding))

Dimensiones: 384


In [37]:
embedding[:10]

array([ 0.09726876,  0.05324143, -0.02505196,  0.00969878, -0.08190185,
        0.01847061, -0.07039174, -0.00430584, -0.04726456,  0.04039362],
      dtype=float32)

## Reflexión técnica

Durante esta etapa se comprobó que el modelo all-MiniLM-L6-v2 genera representaciones vectoriales de 384 dimensiones.

Estos vectores constituyen la base del mecanismo de búsqueda semántica que utilizará SOPHIA en la etapa de Retrieval.

La generación de embeddings separa el significado del texto de su representación original, permitiendo encontrar información relevante incluso cuando la consulta utiliza palabras distintas a las presentes en los documentos.

## Validación del modelo

Se comprobó que el modelo `all-MiniLM-L6-v2` genera un vector de **384 dimensiones** para cualquier texto de entrada.

Aunque la longitud del vector permanece constante, sus valores cambian según el contenido del texto. Esto confirma que el modelo representa el significado semántico de cada texto mediante una representación numérica de tamaño fijo.

Esta propiedad permitirá, en las siguientes etapas del proyecto, comparar consultas y documentos utilizando medidas de similitud semántica en lugar de coincidencias exactas de palabras.

---
## Aprendizajes de la etapa

Durante esta etapa se comprendió:

- Qué es un embedding.
- Por qué los embeddings permiten búsquedas semánticas.
- Cómo cargar un modelo de Sentence Transformers.
- Cómo generar un vector de representación para un texto.
- La diferencia entre un Chunk y su representación vectorial.